collect_app/src/main/java/org/odk/collect/android/widgets/utilities/PrintableHtmlParser.kt

```diff
@@ -0,0 +1,57 @@
+package org.odk.collect.android.widgets.utilities
+
+import android.graphics.Bitmap
+import android.graphics.BitmapFactory
+import android.util.Base64
+import org.jsoup.Jsoup
+import org.jsoup.nodes.Document
+import org.odk.collect.android.utilities.QuestionMediaManager
+import org.odk.collect.qrcode.QRCodeCreator
+import java.io.ByteArrayOutputStream
+
+class PrintableHtmlParser(private val qrCodeCreator: QRCodeCreator) {
+    fun parse(htmlDocument: String, questionMediaManager: QuestionMediaManager): String {
+        val document = Jsoup.parse(htmlDocument)
+
+        return convertQRCodeElementsToImages(document, questionMediaManager).html()
+    }
+
+    private fun convertQRCodeElementsToImages(document: Document, questionMediaManager: QuestionMediaManager): Document {
+        parseImages(document, questionMediaManager)
+        parseQRCodes(document)
+
+        return document
+    }
+
+    private fun parseImages(document: Document, questionMediaManager: QuestionMediaManager) {
+        for (imgElement in document.getElementsByTag("img")) {
+            val file = questionMediaManager.getAnswerFile(imgElement.attributes().get("src"))
+            if (file != null && file.exists()) {
+                val bitmap = BitmapFactory.decodeFile(file.absolutePath)
+                if (bitmap != null) {
+                    val imageData = bitmapToBase64(bitmap)
+                    imgElement.attr("src", "data:image/png;base64,$imageData")
```

Comment:
Do you have to do this? Is there no way to set the working directory to the media dir when the HTML gets rendered? I don't think it's a big deal either way but would be nice not to have to process big images this way if possible.

Reply:
Right it is possible. Fixed.